### PR TITLE
Optimize config store `List`

### PR DESCRIPTION
**Pull request by YonkaFang**
**Please provide a description of this PR:**

This PR aims to optimize the memory allocation overhead in the List method of configstore. Considering the extensive use of this method in the Istio codebase, which constitutes a critical code path, optimizing it holds significant value.

Specific changes include:

1. Optimizing memory config store `List` method to pre-allocate a slice of appropriate size based on the actual returned data, avoiding multiple slice expansions caused by repeated `append`s.

2. Implementing similar optimizations for the aggregate config store.

3. Adding a `BenchmarkStoreList` test case to observe the actual impact of the optimization.

Optimization results:

```sh
# before
BenchmarkStoreList-8           7071        147289 ns/op      786132 B/op         12 allocs/op

# after
BenchmarkStoreList-8          22273         52986 ns/op      237568 B/op          1 allocs/op
```

### CODE REVIEW

File: pilot/pkg/config/aggregate/config.go [hunk 1]
```diff
@@ -107,21 +107,34 @@ func (cr *store) Get(typ config.GroupVersionKind, name, namespace string) *confi
...
+		curConfigs := store.List(typ, namespace)
+		storeConfigs = append(storeConfigs, curConfigs)
+		configCnt += len(curConfigs)
+	}
+
+	configs = make([]config.Config, 0, configCnt)
+	// Used to remove duplicated config
+	configMap := sets.NewWithLength[types.NamespacedName](configCnt)
+	for _, curConfigs := range storeConfigs {
+		for _, cfg := range curConfigs {
```
Comment (hzxuzhonghu) [inline pilot/pkg/config/aggregate/config.go:131]: we can use slices.FilterInplace here to reduce one loop

Comment (YonkaFang) [reply pilot/pkg/config/aggregate/config.go]: done

File: pilot/pkg/config/aggregate/config.go [hunk 2]
```diff
@@ -107,21 +108,33 @@ func (cr *store) Get(typ config.GroupVersionKind, name, namespace string) *confi
...
 	}
+
+	configs = make([]config.Config, 0, configCnt)
+	// Used to remove duplicated config
+	configMap := sets.NewWithLength[types.NamespacedName](configCnt)
+	for _, curConfigs := range storeConfigs {
+		configs = append(configs, curConfigs...)
+	}
+	configs = slices.FilterInPlace[config.Config](configs, func(cfg config.Config) bool {
+		return !configMap.InsertContains(cfg.NamespacedName())
```
Comment (wulianglongrd) [inline pilot/pkg/config/aggregate/config.go:135]: It is generally recommended to use stateless functions, but `configMap` is changed here.

Comment (YonkaFang) [reply pilot/pkg/config/aggregate/config.go:135]: I understand that if this is a concurrent streaming process, stateful functions might pose concurrency issues. However, considering that this function is encapsulated by Istio and additional "concurrency" in Go is rare, this should not be a very strict constraint.

Alternatively, what would you suggest?

Comment (wulianglongrd) [reply pilot/pkg/config/aggregate/config.go:135]: I'm not sure, maybe like this

https://github.com/istio/istio/blob/705b68aba020229ba90b5032933461469b9a0846/pilot/pkg/networking/core/v1alpha3/httproute.go#L569

Comment (YonkaFang) [reply pilot/pkg/config/aggregate/config.go:135]: u can see the original version https://github.com/istio/istio/compare/228d42007492798952db5b5bcd7b79f0fde5df51..fc706115770aed6055efdb876c0be8c2c673acaf .

Indeed, we have achieved cleaner and more readable code through some encapsulated syntactic sugar, but we have also lost some control over details. It's a trade-off. Following the suggestion above, I have switched to this version.

Comment (wulianglongrd) [reply pilot/pkg/config/aggregate/config.go:135]: Thanks, got it.

**Review by hzxuzhonghu — APPROVED** [review]
LGTM

File: pilot/pkg/config/aggregate/config.go (hunk 2 above)
Comment (howardjohn) [reply pilot/pkg/config/aggregate/config.go:135]: IMO this is invalid. It relies on the assumption that Filter is called for each element in the slice once. This is probably true, but there is no guarantee. It should be stateless.

Why do we append then filter anyways? is it faster than filtering when we append like we did before?

**Review by howardjohn — COMMENTED** [review]
Nice improvement

**Pull request merged**